vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,300 @@
+#include "conditional_contextual_bandit.h"
+
+#include "example.h"
+#include "global_data.h"
+
+#include <numeric>
+
+bool CCB::ec_is_example_header(example& ec)
+{
+  return ec.l.conditional_contextual_bandit.type == example_type::shared;
+}
+
+void CCB::print_update(vw& all, bool is_test, example& ec, multi_ex* ec_seq, bool action_scores)
+{
+  // TODO: Implement for CCB
+  throw std::runtime_error("CCB:print_update not implemented");
+}
+
+template <typename T>
+T read_object(io_buf& cache)
+{
+  char* c;
+  size_t next_read_size = sizeof(T);
+  if (cache.buf_read(c, next_read_size) < next_read_size)
+    THROW("Failed to read CCB label cache");
+  return *reinterpret_cast<T*>(c);
+}
+
+namespace CCB {
+  size_t read_cached_label(shared_data*, void* v, io_buf& cache)
+  {
+    CCB::label* ld = static_cast<CCB::label*>(v);
+    size_t read_count = 0;
+
+    ld->type = static_cast<example_type>(read_object<uint8_t>(cache));
+    read_count += sizeof(uint8_t);
+
+    bool is_outcome_present = read_object<bool>(cache);
+    read_count += sizeof(bool);
+
+    if (is_outcome_present)
+    {
+      ld->outcome = new CCB::conditional_contexual_bandit_outcome();
+      ld->outcome->probabilities = v_init<ACTION_SCORE::action_score>();
+
+      ld->outcome->cost = read_object<float>(cache);
+      read_count += sizeof(float);
+      auto size_probs = read_object<size_t>(cache);
+      read_count += sizeof(size_t);
+
+      for(auto i = 0; i < size_probs; i++)
+      {
+        ld->outcome->probabilities.push_back(read_object<ACTION_SCORE::action_score>(cache));
+        read_count += sizeof(ACTION_SCORE::action_score);
+      }
+    }
+
+    auto size_includes = read_object<size_t>(cache);
+    read_count += sizeof(size_t);
+
+    for (auto i = 0; i < size_includes; i++)
+    {
+      ld->explicit_included_actions.push_back(read_object<uint32_t>(cache));
+      read_count += sizeof(uint32_t);
+    }
+
+    return read_count;
+  }
+
+  float ccb_weight(void*)
+  {
+    return 1.;
+  }
+
+  void cache_label(void* v, io_buf& cache)
+  {
+    char* c;
+    CCB::label* ld = static_cast<CCB::label*>(v);
+    size_t size =
+      sizeof(uint8_t) // type
+      + sizeof(bool) // outcome exists?
+      + (ld->outcome == nullptr ? 0 :
+        sizeof(float) // cost
+        + sizeof(size_t) // probabilities size
```

Comment:
We should avoid read/write of objects of nonfixed size to preserve compatibility across 32/64 bit platforms.

Reply:
I am not sure how we can avoid this here?

Reply:
Arrange to use uint32_t or uint64_t.

Reply:
Should we just enforce that uint32_t is used here then?

Reply:
Yes, make the relevant variables have a fixed length.

Reply:
More generally, it seems better to use sizeof(variable_name) than sizeof(typename) since that better documents why a particular size is needed.